packages/@okta/vuepress-site/code/java/spring_security_saml/index.md

```diff
@@ -7,217 +7,372 @@ excerpt: >-
-```
+1. Create a `src/main/resources/application.yml` file to contain your metadata URI.
 
-**7. Start the Application**
+   ```yaml
+   spring:
+     security:
+       saml2:
+         relyingparty:
+           registration:
+             okta:
+               assertingparty:
+                 metadata-uri: <your-metadata-uri>
+   ```
 
-Load the Spring SAML application by opening this link: `http://localhost:8080/spring-security-saml2-sample/saml/discovery?entityID=http%3A%2F%2Flocalhost%3A8080%2Fspring-security-saml2-sample%2Fsaml%2Fmetadata&returnIDParam=idp`
+### Run the app and authenticate
 
-> **Note:** Links in the app aren't functional yet because we haven't configured any IDPs. Full app functionality is completed after the **Configure Spring Security SAML to work with Okta** section.
+1. Run your Spring Boot app. You can do this from your IDE, or as follows using the command line:
 
-Here's what it should look like: ![App Running](/img/spring-security-saml-intro.png "App Running")
+   ```shell
+   ./gradlew bootRun
+   ```
 
-## Configure Okta to work with Spring Security SAML
+1. Open `http://localhost:8080` in your favorite browser and sign in with the credentials you used to create your account. You should see a successful result in your browser.
+
+   <div class="three-quarter border">
+
+   ![Login success](/img/saml/spring-security-saml-login.png)
+
+   </div>
+
+If you try to sign out, it won't work. You'll fix that in the next section.
+
+### Add a sign out feature
+
+Spring Security's SAML support has a [sign out feature](https://docs.spring.io/spring-security/reference/servlet/saml2/logout.html) that requires configuration.
+
+1. Edit your application on Okta and navigate to **General** > **SAML Settings** > **Edit**.
+
+1. Continue to the **Configure SAML** step and **Show Advanced Settings**. Before you can enable single logout, you'll have to create and upload a certificate to sign the outgoing logout request.
+
+1. You can create a private key and certificate using OpenSSL. Answer at least one of the questions with a value, and it should work.
+
+   ```shell
+   openssl req -newkey rsa:2048 -nodes -keyout local.key -x509 -days 365 -out local.crt
+   ```
+
+1. Copy the generated files to your app's `src/main/resources` directory. Configure `signing` and `singlelogout` in `application.yml`:
+
+   ```yaml
+   spring:
+     security:
+       saml2:
+         relyingparty:
+           registration:
+             okta:
+               signing:
+                 credentials:
+                   - private-key-location: classpath:local.key
+                     certificate-location: classpath:local.crt
+               singlelogout:
+                 binding: POST
+                 response-url: "{baseUrl}/logout/saml2/slo"
+               assertingparty:
+                 metadata-uri: <your-metadata-uri>
+   ```
+
+1. Upload the `local.crt` to your Okta app. Select *Enable Single Logout* and use the following values:
```

Comment:
What do you mean by "Upload" the file to your okta app? You've already copied it and the key file to the app's codebase in step 4, so do you mean something else? If so, how do I do this?

Also, I don't see an option for *Enable Single Logout* under Advanced Settings or anywhere on this page. Has the UI changed?

Reply:
There's a file upload button when you enable Single Logout for a SAML App. It's under Advanced Settings.

<img width="700" alt="Screen Shot 2022-08-18 at 11 54 56" src="https://user-images.githubusercontent.com/17892/185462314-ce1680cb-da5e-4204-8f21-0ff8b84aae85.png">

Reply:
Ah I see. I had missed the checkbox to make the other three fields appear. Here's how I'd suggest the instructions should read.

Spring Security's SAML support has a sign out feature that requires a private key and certificate. To use it, you'll need to:

1. Create a private key and certificate to sign the outgoing sign-out request using OpenSSL.

   ```shell
   openssl req -newkey rsa:2048 -nodes -keyout local.key -x509 -days 365 -out local.crt
   ```

2. Copy the generated `local.crt` and `local.key` files to your app's `src/main/resources` directory.
3. Update your `application.yml` configuration file to refer to the new certificate files.

   ```yaml
   spring:
     security:
       saml2:
         relyingparty:
           registration:
             okta:
               signing:
                 credentials:
                   - private-key-location: classpath:local.key
                     certificate-location: classpath:local.crt
               singlelogout:
                 binding: POST
                 response-url: "http:/localhost:8080/logout/saml2/slo"
               assertingparty:
                 metadata-uri: <your-metadata-uri>
   ```

4. Then add this certificate to your app integration.

   1. Open the **Admin Console** for your org.
   2. Choose **Applications** > **Applications**.
   3. Click on the name of your SAML app integration.
   4. Choose the **General** tab, then locate the **SAML Settings** section, and click **Edit**.
   5. Click **Next**.
   6. In the **SAML Settings** section
      - Click **Show Advanced Settings**.
      - Select **Allow application to initiate Single Logout** for **Enable Single Logout**.
      - Set **Single Logout URL** to to a URL that is appropriate for your app. For example `http://localhost:8080/logout/saml2/slo`
      - Set **SP Issuer** to a URL that is appropriate for your app. For example `http://localhost:8080/saml2/service-provider-metadata/okta`
      - Click the **Browse** button for **Signature Certificate**, locate the `local.crt` file you created in step 1, and click **Upload Certificate**.
      - Click **Next**.
   7. Click **Finish**.

Reply:
OK, I've made these updates in my latest commit

Reply:
Looks good